File: Tree/treesideview.c
/*
Given a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.
*/

typedef struct queueNode {
	treeNode 	*node;
	int 		level;
};


treeNode *sideView[MAX_TREE_DEPTH];

/* Algorithm:
	1. Take a array of size tree depth. This array will contain side view tree nodes.
	2. The idea is at each level we will enqueue right child first and then left. So it is guranteed that right child will always
	   get processed before left child. 
	3. If at any level we will process right subtree first and then left subtree. 
	4. As we are processing right subtree first, sideView array at each level will always contain the right most child in right subtree.
	5. If at any level right subtree does not exist, left subtree is choosen.
*/
int populateSideView(treeNode *root)
{
	queueNode *node=NULL;
	
	//Queue will have node and its depth.
	enQueue(root, 0);
	
	while ((node = dQueue()) != NULL) {
		if (sideView[node->level] == NULL) {
			sideView[node->level] = node->node;
		}
		if (node->node->right != NULL) { enQueue(node->node->right, node->level+1);}
		if (node->node->left != NULL) { enQueue(node->node->left, node->level+1);}
	}
}

File: Tree/treesideview.py
'''
Given a binary tree, imagine yourself standing on the right side of it, return the values of the nodes you can see ordered from top to bottom.

Example:

Input: [1,2,3,null,5,null,4]
Output: [1, 3, 4]
Explanation:

   1            <---
 /   \
2     3         <---
 \     \
  5     4       <---
'''
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.min = 0
        self.max = 0
        
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = list()
        if root is None: return result
        level_track = dict()
        self.rightSideView_recursive(root, 0, level_track)
        print self.min, self.max
        result = list()
        i= self.min
        while i<=self.max:
            result.append(level_track[i])
            i +=1
        
        return result
    
    def rightSideView_recursive(self, node, level, level_track):
        if node is None: return
        if level not in level_track:
            level_track[level] = node.val
        
        if level < self.min: self.min = level
        elif level > self.max: self.max = level
            
        self.rightSideView_recursive(node.right, level+1, level_track)
        self.rightSideView_recursive(node.left, level+1, level_track)
